File: src/CCVTAC.Console/PostProcessing/Renamer.cs
using System.IO;
using System.Text;
using System.Text.RegularExpressions;
using UserSettings = CCVTAC.FSharp.Settings.UserSettings;

namespace CCVTAC.Console.PostProcessing;

internal static class Renamer
{
    public static void Run(
        UserSettings settings,
        string workingDirectory,
        Printer printer)
    {
        Watch watch = new();

        var workingDirInfo = new DirectoryInfo(workingDirectory);
        var audioFiles = workingDirInfo
            .EnumerateFiles()
            .Where(f => PostProcessor.AudioExtensions.CaseInsensitiveContains(f.Extension))
            .ToImmutableList();

        if (audioFiles.None())
        {
            printer.Warning("No audio files to rename were found.");
            return;
        }

        printer.Debug($"Renaming {audioFiles.Count} audio file(s)...");

        Regex regex;
        List<Match> matches;
        string matchedPatternSummary;

        foreach (FileInfo file in audioFiles)
        {
            var newFileName = settings.RenamePatterns.Aggregate(
                new StringBuilder(file.Name),
                (newNameSb, renamePattern) =>
                {
                    regex = new Regex(renamePattern.RegexPattern);

                    // A match is generated for each instance of the matched substring.
                    matches = regex
                        .Matches(newNameSb.ToString())
                        .Where(m => m.Success)
                        .Reverse() // Avoids index errors.
                        .ToList();

                    if (matches.Count == 0)
                    {
                        return newNameSb;
                    }

                    if (!settings.QuietMode)
                    {
                        matchedPatternSummary = renamePattern.Summary is null
                            ? $"`{renamePattern.RegexPattern}` (no description)"
                            : $"\"{renamePattern.Summary}\"";

                        printer.Debug($"Rename pattern {matchedPatternSummary} matched × {matches.Count}.");
                    }

                    foreach (Match match in matches)
                    {
                        // Delete the matched substring from the filename by index.
                        newNameSb.Remove(match.Index, match.Length);

                        // Generate replacement text to be inserted at the same starting index
                        // using the matches and the replacement patterns from the settings.
                        var replacementText =
                            match.Groups.OfType<Group>()
                                // `Select()` indexing begins at 0, but usable regex matches begin at 1,
                                // so add 1 to both the match group and replacement placeholder indices.
                                .Select((_, i) =>
                                (
                                    SearchFor:   $"%<{i + 1}>s", // Start with placeholder #1 because...
                                    ReplaceWith: match.Groups[i + 1].Value.Trim() // ...we start with regex group #1.
                                ))
                                // Starting with the placeholder text from the settings, replace each
                                // individual placeholder with the correlated match text, then
                                // return the final string.
                                .Aggregate(
                                    new StringBuilder(renamePattern.ReplaceWithPattern),
                                    (workingText, replacementParts) =>
                                        workingText.Replace(
                                            replacementParts.SearchFor,
                                            replacementParts.ReplaceWith),
                                    workingText => workingText.ToString()
                                );

                        newNameSb.Insert(match.Index, replacementText);
                    }

                    return newNameSb;
                },
                newFileNameSb => newFileNameSb.ToString());

            try
            {
                File.Move(
                    file.FullName,
                    Path.Combine(workingDirectory, newFileName)
                        .Normalize(GetNormalizationForm(settings.NormalizationForm)));

                printer.Debug($"• From: \"{file.Name}\"");
                printer.Debug($"    To: \"{newFileName}\"");
            }
            catch (Exception ex)
            {
                printer.Error($"• Error renaming \"{file.Name}\": {ex.Message}");
            }
        }

        printer.Info($"Renaming done in {watch.ElapsedFriendly}.");
    }

    private static NormalizationForm GetNormalizationForm(string form) =>
        form.Trim().ToUpperInvariant() switch
        {
            "D" => NormalizationForm.FormD,
            "KD" => NormalizationForm.FormKD,
            "KC" => NormalizationForm.FormKC,
            _ => NormalizationForm.FormC
        };
}

File: src/CCVTAC.FSharp/PostProcessing/Renamer.fs
namespace CCVTAC.Console.PostProcessing

open CCVTAC.Console
open CCVTAC.Console.IoUtilities
open CCVTAC.Console.Settings.Settings
open System
open System.IO
open System.Text
open System.Text.RegularExpressions
open Startwatch.Library

module Renamer =

    let private toNormalizationForm (form: string) =
        match form.Trim().ToUpperInvariant() with
        | "D"  -> NormalizationForm.FormD
        | "KD" -> NormalizationForm.FormKD
        | "KC" -> NormalizationForm.FormKC
        | _    -> NormalizationForm.FormC

    let updateTextViaPatterns isQuietMode (printer: Printer) (sb: SB) (renamePattern: RenamePattern) =
        let regex = Regex renamePattern.RegexPattern

        let matches =
            regex.Matches(sb.ToString())
            |> Seq.cast<Match>
            |> Seq.filter _.Success
            |> Seq.rev
            |> Seq.toList

        if List.isEmpty matches
        then sb
        else
            if not isQuietMode then
                let patternSummary =
                    if String.hasText renamePattern.Summary then
                        $"\"%s{renamePattern.Summary}\""
                    else
                        $"`%s{renamePattern.RegexPattern}` (no description)"

                printer.Debug $"Rename pattern %s{patternSummary} matched × %d{matches.Length}."

            for m in matches do
                sb.Remove(m.Index, m.Length) |> ignore

                // Build replacement text by replacing %<n> placeholders with group captures.
                let replacementText =
                    m.Groups
                    |> Seq.cast<Group>
                    |> Seq.mapi (fun i _ ->
                        let searchFor = sprintf "%%<%d>s" (i + 1)
                        let replaceWith =
                            // Group 0 is the entire match, so we only want groups starting at 1.
                            if i + 1 < m.Groups.Count
                            then m.Groups[i + 1].Value.Trim()
                            else String.Empty
                        (searchFor, replaceWith))
                    |> Seq.fold (fun (sb': SB) -> sb'.Replace) (SB renamePattern.ReplaceWithPattern)
                    |> _.ToString()

                sb.Insert(m.Index, replacementText) |> ignore

            sb

    let run userSettings workingDirectory (printer: Printer) : unit =
        let watch = Watch()
        let workingDirInfo = DirectoryInfo workingDirectory

        let audioFiles =
            workingDirInfo.EnumerateFiles()
            |> Seq.filter (fun f -> List.caseInsensitiveContains f.Extension Files.audioFileExtensions)
            |> List.ofSeq

        if List.isEmpty audioFiles then
            printer.Warning "No audio files to rename were found."
        else
            printer.Debug $"""Renaming %s{String.fileLabel (Some "audio") audioFiles.Length}..."""

            for audioFile in audioFiles do
                let newFileName =
                    userSettings.RenamePatterns
                    |> List.fold
                        (fun (sb: SB) -> updateTextViaPatterns userSettings.QuietMode printer sb)
                        (SB audioFile.Name)
                    |> _.ToString()

                try
                    let destinationPath =
                        Path.Combine(workingDirectory, newFileName)
                        |> _.Normalize(toNormalizationForm userSettings.NormalizationForm)

                    File.Move(audioFile.FullName, destinationPath)

                    printer.Debug $"• From: \"%s{audioFile.Name}\""
                    printer.Debug $"    To: \"%s{newFileName}\""
                with ex ->
                    printer.Error $"• Error renaming \"%s{audioFile.Name}\": %s{ex.Message}"

            printer.Info $"Renaming done in %s{watch.ElapsedFriendly}."
